lighthouse-core/gather/fetcher.js

```diff
@@ -0,0 +1,177 @@
+    this._onRequestPausedHandlers = new Map();
+    this._onRequestPaused = this._onRequestPaused.bind(this);
+  }
+
+  /**
+   * The Fetch domain accepts patterns for controlling what requests are intercepted, but we
+   * enable the domain for all patterns and filter events at a lower level to support multiple
+   * concurrent usages. Reasons for this:
+   *
+   * 1) only one set of patterns may be applied for the entire domain.
+   * 2) every request that matches the patterns are paused and only resumes when certain Fetch
+   *    commands are sent. So a listener of the `Fetch.requestPaused` event must either handle
+   *    the requests it cares about, or explicitly allow them to continue.
+   * 3) if multiple commands to continue the same request are sent, protocol errors occur.
+   *
+   * So instead we have one global `Fetch.enable` / `Fetch.requestPaused` pair, and allow specific
+   * urls to be intercepted via `driver.setOnRequestPausedHandler`.
+   */
+  async enableRequestInterception() {
+    await this.driver.sendCommand('Fetch.enable', {
+      patterns: [{requestStage: 'Request'}, {requestStage: 'Response'}],
+    });
+    await this.driver.on('Fetch.requestPaused', this._onRequestPaused);
+  }
+
+  /**
+   * @param {string} url
+   * @param {(event: LH.Crdp.Fetch.RequestPausedEvent) => void} handler
+   */
+  async setOnRequestPausedHandler(url, handler) {
+    this._onRequestPausedHandlers.set(url, handler);
+  }
+
+  /**
+   * @param {LH.Crdp.Fetch.RequestPausedEvent} event
+   */
+  async _onRequestPaused(event) {
+    const handler = this._onRequestPausedHandlers.get(event.request.url);
+    if (handler) {
+      await handler(event);
+    } else {
+      // Nothing cares about this URL, so continue.
+      await this.driver.sendCommand('Fetch.continueRequest', {requestId: event.requestId});
+    }
+  }
+
+  async disableRequestInterception() {
+    await this.driver.off('Fetch.requestPaused', this._onRequestPaused);
+    await this.driver.sendCommand('Fetch.disable');
+    this._onRequestPausedHandlers.clear();
+  }
+
+  /**
+   * Requires that `driver.enableRequestInterception` has been called.
+   *
+   * Fetches any resource in a way that circumvents CORS.
+   *
+   * @param {string} url
+   * @param {number} timeoutInMs
+   * @return {Promise<string>}
+   */
+  async fetchResource(url, timeoutInMs = 500) {
+    if (!this.driver.isDomainEnabled('Fetch')) {
+      throw new Error('Must call `enableRequestInterception` before using fetchResource');
+    }
+
+    /** @type {Promise<string>} */
+    const requestInterceptionPromise = new Promise((resolve, reject) => {
+      this.setOnRequestPausedHandler(url, async (event) => {
+        const {requestId, responseStatusCode} = event;
+
+        // The first requestPaused event is for the request stage. Continue it.
+        if (!responseStatusCode) {
+          // Remove same-site cookies so we aren't buying stuff on Amazon.
+          const headers = [];
```

Comment:
seems like we don't want to be removing all the headers right, just replacing the Cookie one?

AFAICT that's what this API will do though I might be misunderstanding something 
https://cs.chromium.org/chromium/src/content/browser/devtools/devtools_url_loader_interceptor.cc?sq=package:chromium&g=0&l=954-965

Reply:
you're right :)